### ISSUE TITLE
精度问题

### ISSUE THREAD

**Issue by gubeo** [kt4ngw/ICC-2024#1]

**Comment by kt4ngw** [comment]
请问具体是什么问题？跑一次可能精度没有某些对比算法高，但多次平均的准确率应该是大于基准的，这是因为100中选择10，FL本身就具有随机性。

**Comment by gubeo** [comment]
> 请问是什么具体问题？跑一次可能精度没有某些对比算法高，但多次平均的准确率应该是大于基准的，这是因为100中选择10，FL本身就具有随机性。
propose策略下每次精度还是挺高的，但是去平均后画出来的图就很低
![Figure_2](https://github.com/user-attachments/assets/e3ec80e7-1a0c-4f70-b9cd-69f3c68216b3)

**Comment by kt4ngw** [comment]
你检查一下是不是某些随机种子初始化，模型陷入局部次优了？

**Comment by gubeo** [comment]
请问具体在哪里找，怎么判断是否陷入局部最优

**Comment by kt4ngw** [comment]
通过观察，propose的某一个随机种子，到10-20轮，正确率一直是20%不动了，可能就是陷入局部最优了。
神经网络偶尔会这样。
如果按你说的①propose策略下每次精度还是挺高的，但是平均画图后出现这个效果，②以及propose没有上述的情况，那可能就是你画图的代码出现了一些问题。

**Comment by gubeo** [comment]
感谢回答，经观察，propose策略下全局精度在10-20轮左右精度在80%-95%，可能是画图问题，但是画的图也是根据数据呀，所以不明白为什么数据很好但是图不好，请问方便加个联系方式进行讨论吗

**Comment by kt4ngw** [comment]
你可以将绘图代码发到这里。联系方式你可以发我邮箱，这里不方便发联系方式。

**Comment by gubeo** [comment]
import json
import os
# 构建文件路径
import matplotlib.pyplot as plt
import numpy as np
folder_path = os.path.join('D:\\Python1\\Z\\ICC-2024-main', "result\\mnist_dir_0.1")
# file_path_us = os.path.join(folder_path, "fedavg_mnist_cnn__wn10_tn100_sd9_lr0.001_ne100_bs32/metrics.json")
# file_path_pro = os.path.join(folder_path, "propose_mnist_cnn__wn10_tn100_sd9_lr0.001_ne100_bs32/metrics.json")


print(f'Folder path: {os.path.abspath(folder_path)}')

pro_acc = [0] * 302
pro_loss = [0] * 302
pro_energy = [0] * 302
pro_delay = [0] * 302
for i in range(0, 5):
    # s = 'propose_mnist_cnn_sgd__wn10_tn100_sd' + str(i) + '_lr0.1_ne301_bs32/metrics.json'
    s = f'propose_mnist_cnn_sgd__wn10_tn100_sd' + str(i) + '_lr0.1_ne301_bs32\\metrics.json'

file_path_us = os.path.join(folder_path, s)
print(file_path_us)

try:
    with open(file_path_us, 'r') as json_file:
        data = json.load(json_file)
        # 在这里，你可以处理数据
except FileNotFoundError:
    print(f'File not found: {file_path_us}')

with open(file_path_us, 'r') as json_file:
    data = json.load(json_file)
    for j in range(len(data['acc_on_g_test_data'])):
        pro_acc[j] += 1 / 5 * data['acc_on_g_test_data'][j]
        pro_loss[j] += 1 / 5 * data['loss_on_g_test_data'][j]
        pro_energy[j] += 1 / 5 * data['accumulation_energy'][j]
        pro_delay[j] += 1 / 5 * data['accumulation_delay'][j]


dict_pro_100 = {}
for i in range(0, len(pro_acc)):
    l = []
    l.append(i)
    l.append(pro_loss[i])
    l.append(pro_acc[i])
    l.append(pro_energy[i - 1])
    # l.append(pro_energy[i])
    dict_pro_100[str(i)] = l


us_acc = [0] * 302
us_loss = [0] * 302
us_energy = [0] * 302
us_delay = [0] * 302

for i in range(0, 5):
    s = f'fedavg_mnist_cnn_sgd__wn10_tn100_sd' + str(i) + '_lr0.1_ne301_bs32\\metrics.json'
    file_path_us = os.path.join(folder_path, s)
    with open(file_path_us, 'r') as json_file:
        data = json.load(json_file)

        for j in range(len(data['acc_on_g_test_data'])):
            us_acc[j] += 1 / 5 * data['acc_on_g_test_data'][j]
            us_loss[j] += 1 / 5 * data['loss_on_g_test_data'][j]
            us_energy[j] += 1 / 5    * data['accumulation_energy'][j]
            us_delay[j]  += 1 / 5 * data['accumulation_delay'][j]

dict_us_100 = {}
for i in range(0, len(pro_acc)):
    l = []
    l.append(i)
    l.append(us_loss[i])
    l.append(us_acc[i])
    l.append(us_energy[i - 1])
    # l.append(us_energy[i ])
    dict_us_100[str(i)] = l


pro2_acc = [0] * 302
pro2_loss = [0] * 302
pro2_energy = [0] * 302
pro2_delay = [0] * 302

for i in range(0, 5):
    s = 'propose2_mnist_cnn_sgd__wn10_tn100_sd' + str(i) + '_lr0.1_ne301_bs32/metrics.json'
    file_path_us = os.path.join(folder_path, s)
    with open(file_path_us, 'r') as json_file:
        data = json.load(json_file)

        for j in range(len(data['acc_on_g_test_data'])):
            pro2_acc[j] += 1 / 5 * data['acc_on_g_test_data'][j]
            pro2_loss[j] += 1 / 5 * data['loss_on_g_test_data'][j]
            pro2_energy[j] += 1 / 5 * data['accumulation_energy'][j]
            pro2_delay[j]  += 1 / 5 * data['accumulation_delay'][j]
dict_pro2_100 = {}
for i in range(0, len(pro_acc)):
    l = []
    l.append(i)
    l.append(pro2_loss[i])
    l.append(pro2_acc[i])
    l.append(pro2_energy[i - 1])
    # l.append(pro2_energy[i])
    dict_pro2_100[str(i)] = l


nws_acc = [0] * 302
nws_loss = [0] * 302
nws_energy = [0] * 302
nws_delay = [0] * 302

for i in range(0, 5):
    s = 'nws_mnist_cnn_sgd__wn10_tn100_sd' + str(i) + '_lr0.1_ne301_bs32/metrics.json'
    file_path_us = os.path.join(folder_path, s)
    with open(file_path_us, 'r') as json_file:
        data = json.load(json_file)

        for j in range(len(data['acc_on_g_test_data'])):
            nws_acc[j] += 1 / 5* data['acc_on_g_test_data'][j]
            nws_loss[j] += 1 /5 * data['loss_on_g_test_data'][j]
            nws_energy[j] += 1 / 5  * data['accumulation_energy'][j]
            nws_delay[j]  += 1 / 5* data['accumulation_delay'][j]

dict_nws_100 = {}
for i in range(0, len(pro_acc)):
    l = []
    l.append(i)
    l.append(nws_loss[i])
    l.append(nws_acc[i])
    l.append(nws_energy[i - 1])
    # l.append(nws_energy[i])
    dict_nws_100[str(i)] = l


if __name__ == "__main__":
    import pandas as pd

    # 创建一个包含数据的 DataFrame
    data = {
        'pro_loss': pro_loss[1:301],
        'nws_loss': nws_loss[1:301],
        'us_loss': us_loss[1:301],
        'pro2_loss': pro2_loss[1:301]
    }

    df = pd.DataFrame(data)

    # 将数据保存到 Excel 文件
    df.to_excel('loss_data.xlsx', index=False)
    # 创建一个包含数据的 DataFrame
    data = {
        'pro_accu': pro_acc[1:301],
        'nws_acc': nws_acc[1:301],
        'us_acc': us_acc[1:301],
        'pro2_acc': pro2_acc[1:301]
    }

    df = pd.DataFrame(data)
    # 将数据保存到 Excel 文件
    df.to_excel('acc_data.xlsx', index=False)
   

    round = [_ for _ in range(301)]
    plt.figure(figsize=(7, 7))
    # plt.plot(round[1:201:4], us_loss[1:201:4], color = '#4F9DA6', linestyle='-', label='baseline1')
    # plt.plot(round[1:201:4], pro_loss[1:201:4],  color = '#1A0841', linewidth=2, linestyle='-', label='proposed')
    # plt.plot(round[1:201:4], pro2_loss[1:201:4],  color = 'r', linewidth=2, linestyle='-', label='proposed2')
    plt.plot(round[:300:], pro_loss[1:301:], color='#403990',  linestyle='-', label='Proposed')
    plt.plot(round[:300:], nws_loss[1:301:], color='#F46F43',  linestyle='-', label='D-WS')
    plt.plot(round[:300:], us_loss[1:301:], color='#80A6E2', linestyle='-', label='US')
    plt.plot(round[:300:], pro2_loss[1:301:], color='#FBDD85',  linestyle='-', label='CCPS')

    # # 添加水平直线   '
    y_position = 0.1  # 调整为你想要的 y 轴位置
    plt.axhline(y=y_position, color='red', linestyle='-', alpha=0.35)
    plt.xlabel('Rounds', fontsize=10, )
    plt.ylabel('Loss', fontsize=10, )
    plt.xlim(0, 300)
    plt.ylim(0, 0.6)
    plt.xticks(fontsize=10)
    plt.yticks(fontsize=10)
    plt.legend(fontsize=10)

    plt.figure(figsize=(7, 7))

    plt.plot(round[:300:], pro_acc[1:301:], color='#403990',  linestyle='-', label='Proposed')
    plt.plot(round[:300:], nws_acc[1:301:], color='#F46F43',  linestyle='-', label='D-WS')
    plt.plot(round[:300:], us_acc[1:301:], color='#80A6E2', linestyle='-', label='US')
    plt.plot(round[:300:], pro2_acc[1:301:], color='#FBDD85',  linestyle='-', label='CCPS')
    plt.xlim(0, 300)
    plt.ylim(0, 1)
    plt.yticks(np.arange(0, 1.01, 0.1), fontsize=35)
    plt.xlabel('Rounds', fontsize=10, )
    plt.ylabel('Accuracy', fontsize=10, )
    plt.xticks(fontsize=10)
    plt.yticks(fontsize=10)
    plt.legend(fontsize=10)

    plt.show()

请帮忙检查一下

**Comment by kt4ngw** [comment]
这里可以发文件吗？代码并不能看出缩进

**Comment by kt4ngw** [comment]
你上传一下文件

**Comment by gubeo** [comment]
好像.py不能发，我发你邮箱加个联系方式吧，方便讨论

**Comment by kt4ngw** [comment]
ok
